docs/proposals/20201019-prometheus-rules-in-observatorium-api.md

```diff
@@ -0,0 +1,94 @@
+
+* **Owners:**:
+  * `@squat` / `@ianbillett`
+
+* **Related Tickets:**
+  * `<JIRA, GH Issues>`
+
+* **Other docs:**
+  * `<Links…>`
+
+## TL;DR
+
+We propose implementing a multi-tenant API that allows users to create, read, update and delete Prometheus alerting and recording rules.
+
+## Why
+
+The single biggest source of frustration from internal Red Hat users of Observatorium right now is the time it takes for changes to their rules to appear in our production infrastructure.
+
+### Pitfalls of the current solution
+
+#### Implicit deploy cycle dependency
+
+Currently, the recording rules deployed into Observatorium are configured in the same repository as the configuration for the service itself. This implicitly ties the deploy-cycle of Prometheus rules to the deploy-cycle of Observatorium itself. If a new set of rules needs to be deployed, our team is required to roll-out our production infrastructure.
+
+As the number of tenants we support and users we serve increases, the team responsible for the Observatorium installation is required to roll out production more frequently in order to satisfy user requests. This becomes an increasing impediment to user experience as the size of Observatorium increases.
+
+#### Shared configuration repositories
+
+For rules to be rolled out, users need the ability to raise pull requests against the Observatorium repositories that contain our source configuration. This works fine when the two team co-exist within the same organisation, but this becomes a blocking issue when the two teams share a much more restricted (i.e. inter-company) trust boundary.
+
+This is not a blocker for Red Hat's internal offering, however this is not inline with Observatorium's stated goal (TODO: insert link) of offering a SaaS-like monitoring solution.
+
+## Goals
+
+* Users can update their Prometheus rules without intervention from the operating team.
+
+## Non-Goals
+
+* Rate-limiting or accounting of user resources.
+* Define or implement the underlying database technology.
+
+## How
+
+We propose to solve the above problem by implementing an API that allows multi-tenant users to create, read, update and delete Prometheus alerting and recording rules.
+
+Thanos Ruler operates on the rule configuration provided by users, this is consumed from a file in Ruler's local data directory. Therefore, this configuration that we ingest via the API needs to be replicated into the local data directory of the Thanos Ruler instances. This is achieved with the [thanos-ruler-syncer](https://github.com/observatorium/thanos-rule-syncer), an application that calls the API above and writes the results into a location that Thanos Rule can access.
+
+This solution must be backwards compatible with rule configuration currently defined in our repository. That is to say, to ease the transition from our current setup to this new API, Thanos Ruler should be able to operate on Rules both from the API and Supplied as configuration.
```

Comment:
Backwards compatibility sounds like a red hat problem rather than an Observatorium concern, no?

Reply:
Backwards compatibility is - but  the real question here is then: should Observatorium support:
* **only** API driven user rules
* **both** API and config-file user rules?

API-only is easier to support, but I would err on the side of supporting both as config-file are what the upstream communities use. WDYT?

Reply:
hm ok so backwards compatibility for the API means not doing anything, right? Because the API is currently totally agnostic to rules. So we are talking about maintaining backwards compatibility in the observatorium jsonnet and how it automatically configures the thanos ruler?
I am not really sure we should be investing effort in maintaining this ad-hoc jsonnet _API_ since it's not formalized or versioned. It was really a hack to have a way of provisioning rules before we had a proper API. The jsonnet interface has been evolving constantly and changes every time we update kube-thanos :p if we want to, we could go the route of ksonnet and version our jsonnet APIs too, but I wouldn't want this to become a bottleneck right now.

I think that if we want to provide a story for administrators to have a read-only, configuration-based approach to managing recording rules for tenants, we should be opinionated and recommend using a rules-file per tenant, e.g. as implemented in the PR, in conjunction with the thanos-rule-syncer. This way the rules are statically provisioned and managed and we can focus on maintaining a single stable API + configuration file syntax.

Reply:
Ok gotcha, slight misunderstanding there - my intention there was your second point which is a much more eloquent way of putting it.